enzyme/Enzyme/EnzymeLogic.cpp

```diff
@@ -2233,8 +2236,9 @@ Function *CreatePrimalAndGradient(
     }
 
     auto st = dyn_cast<StructType>(foundcalled->getReturnType());
-    bool wrongRet = st == nullptr;
-    if (wrongRet || !hasTape) {
+    bool wrongRet = st == nullptr && !foundcalled->getReturnType()->isVoidTy();
+    if (wrongRet) {
+      // if (wrongRet || !hasTape) {
```

Comment:
will we still hit the `!hasTape`?

Reply:
Both of these cases are commented out because we ideally want to have more permissive parsing of custom gradients, but for now fall back to more rigid rules. The commented out code is part of what we may need for more permissive custom gradient parsing.